src/rdkafka_ssl.c

```diff
@@ -1298,6 +1367,51 @@ int rd_kafka_ssl_ctx_init (rd_kafka_t *rk, char *errstr, size_t errstr_size) {
         }
 #endif
 
+#if OPENSSL_VERSION_NUMBER >= 0x10100000
+        if (rk->rk_conf.ssl.engine_location && !rk->rk_conf.ssl.engine) {
+                /* OpenSSL loads an engine as dynamic id and stores it in internal
+                 * list, as per LIST_ADD command below. If engine already exists
+                 * in internal list, it is supposed to be fetched using engine id. 
+                 */
+                rk->rk_conf.ssl.engine = ENGINE_by_id(rk->rk_conf.ssl.engine_id);
+                if (!rk->rk_conf.ssl.engine) {
+                        rk->rk_conf.ssl.engine = ENGINE_by_id("dynamic");
+                        if (!rk->rk_conf.ssl.engine) {
+                                rd_snprintf(errstr, errstr_size, 
+                                            "ENGINE_by_id failed: ");
+                                goto fail;
+                        }
+                }
+
+                if (!ENGINE_ctrl_cmd_string(rk->rk_conf.ssl.engine, "SO_PATH",
+                                            rk->rk_conf.ssl.engine_location, 0)) {
+                        rd_snprintf(errstr, errstr_size, 
+                                    "ENGINE_ctrl_cmd_string SO_PATH failed: ");
+                        goto fail;
+                }
+
+                if (!ENGINE_ctrl_cmd_string(rk->rk_conf.ssl.engine, "LIST_ADD",
+                                            "1", 0)) {
+                        rd_snprintf(errstr, errstr_size,
+                                    "ENGINE_ctrl_cmd_string LIST_ADD failed: ");
+                        goto fail;
+                }
+
+                if (!ENGINE_ctrl_cmd_string(rk->rk_conf.ssl.engine, "LOAD",
+                                            NULL, 0)) {
+                        rd_snprintf(errstr, errstr_size,
+                                    "ENGINE_ctrl_cmd_string LOAD failed: ");
+                        goto fail;
+                }
+
+                if (!ENGINE_init(rk->rk_conf.ssl.engine)) {
```

Comment:
The fail: label should call ENGINE_free(), if non-NULL, right?

And rd_kafka_ssl_cxt_term() should call ENGINE_free() too, right?

Reply:
if ENGINE_init fails then rk->rk_conf.ssl.engine will be null.
call in rd_kafka_ssl_ctx_term makes sense.